examples/custom-ipfs-repo/index.js

```diff
@@ -0,0 +1,77 @@
+'use strict'
+
+const series = require('async/series')
+const IPFS = require('ipfs')
+
+// Create our custom options
+const customRepositoryOptions = {
+
+  /**
+   * Storage Backends are fully customizable. Each backend can be stored in seperate services,
+   * or in a single service. Options can be passed into the datastores via the storageBackendOptions
+   * property shown below.
+   */
+  storageBackends: {
+    root: require('datastore-fs'), // version and config data will be saved here
+    blocks: require('datastore-fs'),
+    keys: require('datastore-fs'),
+    datastore: require('datastore-fs')
+  },
+
+  /**
+   * Storage Backend Options will get passed into the instantiation of their counterpart
+   * in Storage Backends. If you create a custom datastore, this is where you can pass in
+   * custom insantiation. You can see an S3 datastore example at:
+   * https://github.com/ipfs/js-datastore-s3/tree/master/examples/full-s3-repo
+   */
+  storageBackendOptions: {
+    blocks: {
+      sharding: true,
+      extension: '.data'
+    }
+  },
+
+  // false will disable locking, you can also pass in a custom locker
+  locker: false
+}
+
+// Initialize our IPFS node with the custom repo options
+const node = new IPFS({
+  repo: '/tmp/custom-repo/.ipfs',
+  repoOptions: customRepositoryOptions
+})
+
+// Test the new repo by adding and fetching some data
+let fileMultihash
+series([
+  (cb) => node.on('ready', cb),
+  (cb) => node.version((err, version) => {
+    if (err) {
+      return cb(err)
+    }
+
+    console.log('Version:', version.version)
+    cb()
+  }),
+  (cb) => node.files.add({
+    path: 'test-data.txt',
+    content: Buffer.from('We are using a customized repo!')
+  }, (err, filesAdded) => {
+    if (err) {
+      return cb(err)
+    }
+
+    console.log('\nAdded file:', filesAdded[0].path, filesAdded[0].hash)
+    fileMultihash = filesAdded[0].hash
+    cb()
+  }),
```

Comment:
If this were split into steps and you used `async.seq` instead of `async.series`:

```js
(cb) => node.version(cb),
(version, cb) => {
  console.log('Version:', version.version)
  cb()
},
(cb) => node.files.add({
  path: 'test-data.txt',
  content: Buffer.from('We are using a customized repo!')
}, cb),
(filesAdded, cb) => {
  console.log('\nAdded file:', filesAdded[0].path, filesAdded[0].hash)
  cb(null, filesAdded[0].hash)
},
(fileMultiHash, cb) => node.files.cat(fileMultihash, cb),
// etc.
```

You could get rid of the error handling logic so it reads a bit clearer for people trying to follow it. This would remove the need for the `fileMultihash` global, too.

Alternatively, using the promise-based return values would let you do this without a special dependency:

```js
node.on('ready', () => {
  node.version()
    .then(version => console.log('Version:', version.version)
    .then(() => node.files.add({
      path: 'test-data.txt',
      content: Buffer.from('We are using a customized repo!')
    })
    .then(filesAdded => {
      console.log('\nAdded file:', filesAdded[0].path, filesAdded[0].hash)
      return filesAdded[0].hash
    })
    .then(fileMultihash => node.files.cat(fileMultihash))
    // etc.
})
```

Reply:
It might also be good to catch any errors at the end whether you use `series`, `seq`, or promises, just to demonstrate good practice.

Reply:
I think the promise structure is much easier to consume and that's a great point about the error catching. I'll include those changes in the next update I push.